helm-v3/006-repositories.md

```diff
@@ -17,20 +29,25 @@ and could benefit from a capability to push packages from the Helm client.
 
 To support registries there will be a `helm push` command that works in a
 similar manner to `helm fetch`. It will support different systems via the
-URI/URL scheme and provide a default implementation for HTTP(S). Other systems,
-such as S3, can implement an uploader via a plugin for schemes Helm does not
+URI/URL scheme. It will provide a default implementation for HTTP(S) based
+on the ChartMuseum API spec.
+
+Other systems, such as S3, can implement an uploader via a plugin for schemes Helm does not
 provide an implementation for.
 
-> [name=Matt Farina] There very well may be additional work here around
-> registry search and pluggable or otherwise extendable authenication
-> mechanisms.
-> 
-> [name=Adnan Abdulhussein] What would the default HTTP(S) implementation be?
-> [name=Matt Fisher] Adnan, would you mind clarifying your comment a little?
-> 
-> [name=Adnan Abdulhussein] Sorry, the proposal mentions there will be a
-> default implemention for `helm push` to a regular HTTP(S) chart repository.
-> It's unclear to me how this would actually work.
-> 
-> [name=Adnan Abdulhussein] Sounds like this is something that will be fleshed
-> out more in Josh's proposal, so I'll wait for that :)
+The [helm-push plugin](https://github.com/chartmuseum/helm-push) describes potential semantics
+for a `helm push` command.
+Here are some usage examples:
+ ```
+# push .tgz from "helm package"
+helm push mychart-0.1.0.tgz myrepo
+
+# package and push chart directory
+helm push mychart/ myrepo
+
+# override version in Chart.yaml
+helm push mychart/ --version="0.2.0-rc1" myrepo
+
+# push directly to chart repo URL
+helm push mychart/ https://my.chart.repo.com
```

Comment:
Have you considered the following use case?

The charts are stored and isolated in multiple namespaces and the accesses are controlled by the RBAC process. One user can access the charts under the namespaces only and only if enough permissions are granted.

The whole chart repo system only exposes one unified index entry point for the user to use command `helm repo add <repo_name> <http://system_url/root_entry_point>` to add it to the repo list. This unified index entry point will return all the charts under the namespaces which are accessible to the current authenticated&authorized user.

Switching to the `push` case, as the repo is referred by the root index which contains no namespace info, there should be a way to let users specify which namespace they want to push the charts to when pushing chart artifacts.

Something like:
```
helm push mychart/ my_repo_server myspace

helm push mychart/ --namespace myspace my_repo_server
```